activesupport/test/testing/notification_assertions_test.rb

```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
+
+require_relative "../abstract_unit"
+require "active_support/testing/notification_assertions"
+
+module ActiveSupport
+  module Testing
+    class NotificationAssertionsTest < ActiveSupport::TestCase
+      include NotificationAssertions
+
+      def test_assert_notification
+        assert_notification("post.submitted", title: "Cool Post") do
+          ActiveSupport::Notifications.instrument("post.submitted", title: "Cool Post")
+        end
+
+        assert_notification("post.submitted") do # payload omitted
+          ActiveSupport::Notifications.instrument("post.submitted", title: "Cool Post")
+        end
+
+        assert_raises(Minitest::Assertion, match: /No post.submitted notifications were found/) do
+          assert_notification("post.submitted", title: "Cool Post") { nil } # no notifications
+        end
+
+        match = if RUBY_VERSION >= "3.4"
+          /No post.submitted notification with payload {title: "Cool Post"} was found/
+        else
+          /No post.submitted notification with payload {:title=>"Cool Post"} was found/
+        end
+        assert_raises(Minitest::Assertion, match:) do
+          assert_notification("post.submitted", title: "Cool Post") do
+            ActiveSupport::Notifications.instrument("post.submitted", title: "Cooler Post")
+          end
+        end
```

Comment:
Added this Ruby 3.4 conditional support in response to the following new CI failure.
https://buildkite.com/rails/rails/builds/112693#0192948f-8262-43f3-ba5c-bec1da067b09

Reply:
To avoid the `RUBY_VERSION` conditional, you could also do this:

```rb
        match = /No post.submitted notification with payload #{{title: "Cool Post"}} was found/
```
The string interpolation would rely on the current Ruby version's implementation of `Hash#to_s`.

Though at this point, one could wonder is that writing a test or duplicate the app logic? 😅

Reply:
Ah that's true, could just `to_s` the Hash right in the test to get the relevant implementation for whichever Ruby is running. _Is_ reimplementing the actual code to a degree so I generally prefer the current conditional approach though the suggestion is a bit cleaner.

For context, when I ran into the Ruby 3.4 preview 2 CI failure I searched around the Rails codebase and found various other places using Ruby version conditionals. One linked below.

https://github.com/rails/rails/blob/main/actionpack%2Ftest%2Fdispatch%2Fexception_wrapper_test.rb#L115-L119